### PR TITLE
Preliminary implementation of p4c-of: compiling P4 to DDlog programs that control OvS switches

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
With this PR the core compiler P4->OF is mostly functional.
There are no end-to-end tests, but it generates DDlog code that compiles and looks OK for several tests.
Missing features:
- non-exact key matches
- default_action support
- support for some P4 expressions and statements: isValid(), setValid(), setInvalid(), !(negation), !=, digests, ||.
- end-to-end testing using OvS.
The PR commits are relatively self-contained and can be reviewed separately.

### CODE REVIEW

**Comment by mbudiu-vmw** [comment]
BTW: we may decide to merge the ofp4 branch into main too.

**Comment by mbudiu-vmw** [comment]
@debnil I don't expect you can review all files in this code, but some may benefit from your eye.
We should also talk about how some specific features are represented in the generated DDlog.
I am doing my own DDlog relation generation, which does not generate the exact same results as p4info2ddlog.
If you are interested to learn more about the P4C compiler internals I can also help with reading some of this code.

**Comment by debnil** [comment]
@mbudiu-vmw: This was super interesting to read through! I could broadly understand how information flows through the compiler, and what each part is doing.

Let' definitely talk about the generated DDlog and whether we should bridge the gap between `p4info2ddlog` and the code generated here.

Definitely want to learn more about the p4c internals. Let's talk more IRL.

File: ofp4/ddlog.def [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    cstring toString() const override;
+#nodbprint
+}
+
+/// Terms in a rule
+abstract DDlogTerm : IDDlogNode {}
+
+class DDlogAtom : DDlogTerm {
+    IR::ID relation;
+    DDlogExpression expression;
```
Comment (mbudiu-vmw) [inline ofp4/ddlog.def:125]: this is always a TupleExpression.

File: ofp4/ddlog.def [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    cstring toString() const override;
+#nodbprint
+}
+
+class DDlogExpressionTerm : DDlogTerm {
```
Comment (mbudiu-vmw) [inline ofp4/ddlog.def:130]: boolean expressions in general

File: ofp4/ddlog.cpp [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    result += "(";
+    bool first = true;
+    for (auto p : parameters->parameters) {
+        if (!first)
+            result += ", ";
+        first = false;
+        result += p->toString();
+    }
+    result += "): ";
+    result += returnType->toString() + " = ";
```
Comment (mbudiu-vmw) [inline ofp4/ddlog.cpp:124]: use curly braces here

File: ofp4/backend.h [hunk 4]
```diff
@@ -37,6 +39,55 @@ class BackEnd {
...
+    const IR::Parameter* egress_meta_out = nullptr;
+
+    // These correspond directly to the types of the parameters of the ingress/egress blocks
+    const IR::Type_Struct* Headers = nullptr;  // type of ingress_hdr and egress_hdr
+    const IR::Type_Struct* input_metadata_t = nullptr;  // type of ingress_meta_in, egress_meta_in
+    const IR::Type_Struct* M = nullptr;  // type of ingress_meta and egress_meta
+    const IR::Type_Struct* ingress_to_arch_t = nullptr;  // type of ingress_itoa
+    const IR::Type_Struct* output_metadata_t = nullptr;  // type of ingress_meta_out,egress_meta_out
+
+    size_t startIngressId;  // CFG node id of the entry point to ingress
```
Comment (mbudiu-vmw) [inline ofp4/backend.h:72]: these all end up as OF table ids

**Comment by mbudiu-vmw** [comment]
@blp: did we agree to merge this for now?

**Pull request merged**